### PR TITLE
Preliminary support for gaussian model

### PR DESCRIPTION
**Pull request by jolars**
This pull request adds preliminary support for the gaussian model in sgdnet. It seems to work properly in a few basic tests. The setup of the packages is this:

`sgdnet()` is called from R and is designed to mimic the functionality of `glmnet::glmnet()`. The input is remolded to work with the interface from scikit-learn whereafter the C++ function `FitModel()` is called. Checks for sparsity in the feature matrix are made on the R side and are then forwarded to the `FitModel()` function which calls `SagaSolver()`; the latter is templated to work with sparse and dense matrices and then does all the work.

I also added a method for `coef()` and have set up architecture for supporting other objective functions.

I have only really tested the package properly against `Ridge()` from scikitlearn but it provides very similar output in the tests that I have run.

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -20,11 +19,15 @@ URL: https://github.com/jolars/sgdnet
 BugReports: https://github.com/jolars/sgdnet/issues
 RoxygenNote: 6.0.1
 Roxygen: list(markdown = TRUE)
+Depends: R (>= 3.5.0)
```
Comment (michaelweylandt) [inline DESCRIPTION]: Is this a hard dependency or just we're not testing for anything earlier?

File: R/sgdnet.R [hunk 2]
```diff
@@ -0,0 +1,103 @@
+#' Fit a Generalized Linear Model with Elasticnet Regularization
```
Comment (michaelweylandt) [inline R/sgdnet.R]: Elastic Net is two words

File: R/sgdnet.R [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#'
+#' @param x input matrix
+#' @param y response variable
+#' @param family reponse type
+#' @param alpha elasticnet mixing parameter
+#' @param lambda regularization strength
+#' @param intercept whether to fit an intercept or not
+#' @param maxit maximum number of effective passes (epochs)
+#' @param standardize whether to standardize `x` or not -- ignored when
+#'   `intercept == TRUE`.
```
Comment (michaelweylandt) [inline R/sgdnet.R]: Why is this ignored if we have an intercept? We might still want to standardize `X` so that the columns have the same norm (this is often done to put the variables "on the same footing").

File: R/sgdnet.R [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+#' @param thresh tolerance level for termination of the algorithm
+#' @param return_loss whether to compute and return the loss at each outer
+#'   iteration, only added here for debugging purposes.
```
Comment (michaelweylandt) [inline R/sgdnet.R]: Maybe expose this as via the options mechanism if it's only useful for debugging.

File: R/sgdnet.R [hunk 5]
```diff
@@ -0,0 +1,103 @@
+#' Fit a Generalized Linear Model with Elasticnet Regularization
+#'
+#' @param x input matrix
+#' @param y response variable
+#' @param family reponse type
+#' @param alpha elasticnet mixing parameter
```
Comment (michaelweylandt) [inline R/sgdnet.R]: Two words

File: R/sgdnet.R [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+  # The internal optimizer uses a different penalty construction than
+  # glmnet. Convert lambda vallues to match alpha and beta from scikit-learn.
+  alpha_sklearn <- lambda/2*n_samples*(1 - alpha)
+  beta_sklearn <- lambda/2*n_samples*alpha
+
+  # Convert sparse x to dgCMatrix class from package Matrix.
+  if (is_sparse <- inherits(x, "sparseMatrix")) {
+    x <- methods::as(x, "CsparseMatrix")
+    x <- methods::as(x, "dgCMatrix")
```
Comment (michaelweylandt) [inline R/sgdnet.R]: Why the double conversion? Does `as(x, "dgCMatrix")` not automatically work for all `x` classes?

File: R/sgdnet.R [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                  maxit,
+                  thresh,
+                  return_loss)
+
+  variable_names <- colnames(x)
+
+  # Organize return values
+  a0 <- drop(res$a0)
+  names(a0) <- paste0("s", seq_along(a0) - 1L)
+  beta <- Matrix::Matrix(res$beta)
```
Comment (michaelweylandt) [inline R/sgdnet.R:113]: Is the plan to only support scalar `lambda` values or is that only what is implemented so far? If the former, why make this back into a (sparse) matrix?

File: tests/testthat/test-gaussian.R [hunk 8]
```diff
@@ -0,0 +1,20 @@
+context("gaussian regression")
```
Comment (michaelweylandt) [inline tests/testthat/test-gaussian.R:1]: You can do more tests here based on what we know about the solution for the lasso. At a minimum, I'd test: 

i) lambda = 0 ==> get back OLS solution
ii) lambda > lambda_max ==> get back all zero solution
iii) the correct solution conditional on knowing the active set
iv) Closed form solution for ridge regression

Ping me if you need help devising these tests

File: src/utils.h [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  //       is not fit?
+
+  if (fit_intercept) {
+
+    arma::uword n_features = x.n_rows;
+
+    // Center feature matrix
+    x_offset = arma::mean(x, 1);
+    for (arma::uword i = 0; i < n_features; ++i)
+      x.row(i) -= x_offset(i);
```
Comment (michaelweylandt) [inline src/utils.h]: This worries me a bit - if `x` is sparse and large, this will destroy the sparsity pattern (and blow out memory usage).

File: src/utils.h [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+//' @param y_offset a vector of means for each column in y
+//'
+//' @return Nothing. x and y are scaled and centered.
+//' @keywords internal
+template <typename T>
+void Preprocess(T&            x,
+                arma::mat&    y,
+                const bool    normalize,
+                const bool    fit_intercept,
+                arma::vec&    x_offset,
```
Comment (michaelweylandt) [inline src/utils.h]: Can we use a different term than "offset"? 

"Offset" has a statistical meaning (fixed term added to linear predictor in GLM) and IIRC we plan on supporting that sense of "offset" eventually.

File: src/math.h [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+//' Solves \f$ \argmin_{x} 0.5||x - y||^{2} + \alpha ||x||_{1} \f$.
+//'
+//' @param x a double
+//' @param y a double
+//'
+//' @return
+inline double SoftThresholding(const double x,
+                               const double shrinkage) {
+  return std::max(x - shrinkage, 0.0) - std::max(-x - shrinkage, 0.0);
+}
```
Comment (michaelweylandt) [inline src/math.h]: Should we make a "prox" object from which this inherits? That will make it easier to add other proximal operators (SCAD/MCP/group lasso) eventually

File: src/objectives.h [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+namespace sgdnet {
+  class Objective {
+  public:
+    virtual double Loss(const arma::mat& p, const arma::mat& y) = 0;
+    virtual arma::mat Gradient(const arma::mat& p, const arma::mat& y) = 0;
+  };
+
+  class Gaussian : public Objective {
+  public:
+    double Loss(const arma::mat& p, const arma::mat& y) {
```
Comment (michaelweylandt) [inline src/objectives.h]: Can we call this "mu" to match the statistical literature? (mu = conditional mean = E[Y|X])

**Review by michaelweylandt — COMMENTED** [review]
I think the design looks good overall and the use of templating + sparse X support is nice (though there's one spot where I'm concerned you'll loose the sparsity pattern) I had a bit of trouble following the algorithm at one or two points since it doesn't follow the main SAGA paper explicitly. [BTW - Is the lagged update the algorithm from Appendix D?] 

It's probably worth putting in a vignette / technical note which explains the algorithm in painstaking detail before the summer is over. 

For the Gaussian case, we can add some tests based on what we know from the math (+ some noise for the stochastic-ness). We won't be able to do the same for other losses, but it will be a start. 

Needs more docs eventually, but those can be added later.

File: DESCRIPTION (hunk 1 above)
Comment (jolars) [reply DESCRIPTION]: There's no real reason other than that I'm developing on 3.5.0, which makes it convenient. It looks like we can drop to 3.3.0 without any issues. I have no strong feelings about this.

File: R/sgdnet.R (hunk 2 above)
Comment (jolars) [reply R/sgdnet.R]: I adopted this from the glmnet documentation, but I agree.

File: R/sgdnet.R (hunk 3 above)
Comment (jolars) [reply R/sgdnet.R]: Sorry, that's a mistake in the documentation. It is the other way around. No standardization is done when there is no intercept. This is how they do it in scikit-learn's Ridge. I don't know why though, but I just translated it over for now.

File: R/sgdnet.R (hunk 7 above)
Comment (jolars) [reply R/sgdnet.R:113]: This is only what's implemented so far. It is converted because it then matches glmnet, even though we have a dense weights matrix internally. I don't know if glmnet actually uses a sparse implementation for their weights but it might make sense.

File: tests/testthat/test-gaussian.R (hunk 8 above)
Comment (jolars) [reply tests/testthat/test-gaussian.R:1]: I agree, this should be added too.

File: src/utils.h (hunk 10 above)
Comment (jolars) [reply src/utils.h]: Sure. Center? Mean?

File: src/utils.h (hunk 9 above)
Comment (jolars) [reply src/utils.h]: Of course, you are right. So just stick to scaling when features are sparse?

File: src/math.h (hunk 11 above)
Comment (jolars) [reply src/math.h]: Good idea. I have been thinking of something along those lines already.

File: src/objectives.h (hunk 12 above)
Comment (jolars) [reply src/objectives.h]: Hm, is mu really used that often for conditional mean? I suggest just spelling out `prediction` or `conditonal_mean` instead.

File: DESCRIPTION (hunk 1 above)
Comment (tdhock) [reply DESCRIPTION]: would be good to drop this dep if possible, to allow people with older R to use our pkg

File: R/sgdnet.R [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+#' @param x input matrix
+#' @param y response variable
+#' @param family reponse type
+#' @param alpha elastic net mixing parameter
+#' @param lambda regularization strength
+#' @param intercept whether to fit an intercept or not
+#' @param maxit maximum number of effective passes (epochs)
+#' @param standardize whether to standardize `x` or not -- ignored when
+#'   `intercept == FALSE`.
+#' @param thresh tolerance level for termination of the algorithm
```
Comment (tdhock) [inline R/sgdnet.R]: would be nice to have details about how this tolerance is used -- when do we stop? using the difference between gradient norm, objective function values, etc?

File: src/objectives.h (hunk 12 above)
Comment (michaelweylandt) [reply src/objectives.h]: Fine with me -- I'm always happy with more descriptive names. ("p" just has another meaning as number of features) Both names work well, though I'm slightly biased towards "conditional mean" as "prediction" carries a few more connotations.

File: src/utils.h (hunk 10 above)
Comment (michaelweylandt) [reply src/utils.h]: Either works. When we add observation weights eventually, this might need to be a weighted mean (not sure what `glmnet` does), so "center" might be less ambiguous.

File: src/constants.h [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#ifndef SGDNET_CONSTANTS_
+#define SGDNET_CONSTANTS_
+
+namespace sgdnet {
+  constexpr double SMALL =
+    std::pow(std::numeric_limits<double>::epsilon(), 0.75);
```
Comment (tdhock) [inline src/constants.h]: please add a comment about what this is used for and why you need e^0.75 -- its not obvious to me

File: tests/testthat/test-gaussian.R (hunk 8 above)
Comment (tdhock) [reply tests/testthat/test-gaussian.R:1]: agree

File: tests/testthat/test-gaussian.R [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+test_that("gaussian regression with dense features work", {
+  x <- matrix(rnorm(30), 10, 3)
```
Comment (tdhock) [inline tests/testthat/test-gaussian.R]: would be great if you could use real data rather than (or in addition to) simulated data for the tests

**Review by tdhock — APPROVED** [review]
looks good to me, please merge after addressing my minor comments

File: src/utils.h (hunk 9 above)
Comment (michaelweylandt) [reply src/utils.h]: Scaling isn't the issue - centering is. 

For now, I think it's safe to throw an error, but long term we can probably support sparse X + centering with clever linear algebra. For example, the centered X, which I'll call X_c, can be written as X_c = X - m 1^T where m is the vector of column means and 1 is the vector of all ones. Then, e.g., we can replace X_c^Ty by `(X - m 1^T)^Ty = X^Ty - 1m^Ty` throughout the algorithm and work with X_c without ever materializing it. 

I don't think that's the right adjustment for our algorithm, but it should be possible to work something similar out. Once you write-up the technical / algorithmic docs, I'll take a look and try to work something out.

File: R/sgdnet.R (hunk 6 above)
Comment (jolars) [reply R/sgdnet.R]: I borrowed this from glmnet

https://github.com/cran/glmnet/blob/ff79c8962562e278abfaafb98641a7aecf140b23/R/glmnet.R#L91

It seemed strange to me but I figured it was there for a reason. I'll look into it.

File: src/utils.h (hunk 9 above)
Comment (jolars) [reply src/utils.h]: Yes, sorry, that's what I meant. I suggest, however, that we just scale the features when the input is sparse and don't throw any errors.

Comment (michaelweylandt) [reply src/utils.h]: Errr.... I don't love the idea of the standardization being different for dense and sparse X. Even if we document it prominently, I would be really surprised if `sgdnet(X, y)` and `sgdnet(as.matrix(X), y)` gave different results...

File: R/sgdnet.R (hunk 6 above)
Comment (jolars) [reply R/sgdnet.R]: No, there doesn't seem to be any reason for this. I am dropping it.

File: DESCRIPTION (hunk 1 above)
Comment (jolars) [reply DESCRIPTION]: Done

File: src/utils.h (hunk 9 above)
Comment (jolars) [reply src/utils.h]: Well, if I am not mistaken, the only differences that emerge will result from differences in the relative size of the penalty. Since we are rescaling, we are otherwise going to get back the same coefficients. In any case, there are always going to be some discrepancies because of the differences in how the intercept is handled between the dense and sparse implementations.

```r
x <- Matrix::rsparsematrix(10000, 2, density = 0.5)
x[, 2] <- x[, 2]*100
x[x[, 2] != 0, 2] <- x[x[, 2] != 0, 2] + 10000
y <- x[, 1] + x[, 2]/100

set.seed(0)
coef1 <- coef(sgdnet(x, y, alpha = 0, lambda = 0))
coef2 <- coef(sgdnet(as.matrix(x), y, alpha = 0, lambda = 0))
cbind(coef1, coef2)

# 3 x 2 Matrix of class "dgeMatrix"
# s0          s0
# (Intercept) 0.185315314 0.069078897
# 1           1.005611802 0.993138237
# 2           0.009968166 0.009986274
```

It feels like there should be a way around this. glmnet returns equivalent results for sparse and dense matrices alike. I'll look into how they standardize their features in the sparse implementation.

Comment (michaelweylandt) [reply src/utils.h]: Re: glmnet -- They do a variant of the linear algebra trick I described, though it's a bit simpler since it's a coordinate descent method: x_{sc}^Ty = (x/s - m)^Ty = x^Ty/s - my

Can you elaborate a bit more on this?

    there are always going to be some discrepancies because
    of the differences in how the intercept is handled between
    the dense and sparse implementations.

This doesn't seem right at all. If sklearn does not give the same results for an otherwise equivalent sparse and dense X, I'd vote strongly for diverging from their behavior. I would expect `coef1` and `coef2` to be in every way equivalent (other than computational speed) in your example above and I imagine our users will also.

File: R/sgdnet.R [hunk 16]
```diff
@@ -9,19 +25,21 @@
 #' @param maxit maximum number of effective passes (epochs)
 #' @param standardize whether to standardize `x` or not -- ignored when
 #'   `intercept == FALSE`.
-#' @param thresh tolerance level for termination of the algorithm
-#' @param return_loss whether to compute and return the loss at each outer
-#'   iteration, only added here for debugging purposes.
+#' @param thresh tolerance level for termination of the algorithm. The
+#'   algorithm terminates when
+#'   \deqn{
+#'     \frac{\mathrm{change~in~weights}}{\mathrm{weights}} < \mathrm{thresh}.
```
Comment (michaelweylandt) [inline R/sgdnet.R]: I'd vote to call these "coefficients" since weights is already used to refer to i) observation weights; and ii) feature weights. 

It might be clearer to write this as something like

\frac{\|\beta^{(t)} - \beta^{(t-1)}\|_{\infty}}{\|\beta^{(t)}\|_{\infty}} < \mathrm{thresh}

File: src/prox.h [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+//' Base class for proximal operators
+//'
+//' @param x value
+//' @param shrinkage shrinkage
+//'
+//' @noRd
+//' @keywords internal
+class Prox {
+public:
+  virtual double Evaluate(const double x, const double shrinkage) = 0;
```
Comment (michaelweylandt) [inline src/prox.h:15]: Might be cleaner to use `operator()` here instead. Then you get something like

```
Prox prox = SoftThreshold()
x = SoftThreshold(x, lambda); 
```

Comment (jolars) [reply src/prox.h:15]: Hm, I'm not sure how to do this with the abstract base class i am using. I am only passing a pointer to `LaggedUpdates()`.

File: R/sgdnet.R (hunk 16 above)
Comment (jolars) [reply R/sgdnet.R]: Hm, in that case I think we should probably switch from using weights for feature weights altogether and simply stick to coefficients or betas. It would only be confusing to use beta or coefficient here but then use weights in the internal API.

File: src/prox.h (hunk 17 above)
Comment (michaelweylandt) [reply src/prox.h:15]: I'm still a bit unclear on the use of `LaggedUpdates` until I see the math, but this probably isn't important if you have something that already works.

File: R/sgdnet.R (hunk 16 above)
Comment (michaelweylandt) [reply R/sgdnet.R]: I'd agree, but maybe wait until you're done porting the basics from sklearn to make the initial port easy? Once that's done, a big rename should be pretty straightforward.

File: src/utils.h (hunk 9 above)
Comment (jolars) [reply src/utils.h]: [Someone else already did the legwork on this](https://stackoverflow.com/questions/46835181/sklearn-perceptron-behaving-differently-for-sparse-and-dense) but they basically use a slower learning rate for the intercept in the sparse implementation to avoid intercept oscillation.

Comment (michaelweylandt) [reply src/utils.h]: That's interesting - ok, let's table this for now and possibly come back to it later. It's not worth holding up other development right now. 

Can you open an issue so we don't forget about it?

Comment (jolars) [reply src/utils.h]: Yes, good point. It would not be terribly hard to run some simulations and see how this plays out in our implementation.

[The issue has been filed](https://github.com/jolars/sgdnet/issues/2).

**Comment by jolars** [comment]
I am merging this one now. I think that most of the issues should be solved, apart from constructing a test for the active set as you suggested @michaelweylandt
